From 5fc43e72fca993bf89352fa25532bd3546dc7d58 Mon Sep 17 00:00:00 2001
From: sharan <sharan.yagneswar@mindmusiclabs.com>
Date: Mon, 9 Dec 2019 22:13:54 +0000
Subject: [PATCH] Made clk-bcm2835-aux irqchip ipipe aware

---
 drivers/clk/bcm/clk-bcm2835-aux.c | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/drivers/clk/bcm/clk-bcm2835-aux.c b/drivers/clk/bcm/clk-bcm2835-aux.c
index 41e0702fae46..666decedfc69 100644
--- a/drivers/clk/bcm/clk-bcm2835-aux.c
+++ b/drivers/clk/bcm/clk-bcm2835-aux.c
@@ -104,6 +104,11 @@ static struct irq_chip bcm2835_auxirq_chip = {
 	.name = "bcm2835-auxirq",
 	.irq_mask = bcm2835_auxirq_mask,
 	.irq_unmask = bcm2835_auxirq_unmask,
+#ifdef CONFIG_IPIPE
+	.irq_hold = bcm2835_auxirq_mask,
+	.irq_release = bcm2835_auxirq_unmask,
+#endif
+	.flags = IRQCHIP_PIPELINE_SAFE,
 };
 
 static const struct irq_domain_ops bcm2835_auxirq_ops = {
-- 
2.17.1

