From 2be8ac377b774c1e845b52b28181a4ffcdccca40 Mon Sep 17 00:00:00 2001
From: kknitin <nitin.kulkarni@mindmusiclabs.com>
Date: Thu, 7 Nov 2019 16:24:29 +0000
Subject: [PATCH] fix for lan7xx driver not being ipipe safe

---
 drivers/net/usb/lan78xx.c | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/drivers/net/usb/lan78xx.c b/drivers/net/usb/lan78xx.c
index 8761bc74cd5a..a994906c8e59 100644
--- a/drivers/net/usb/lan78xx.c
+++ b/drivers/net/usb/lan78xx.c
@@ -1930,8 +1930,13 @@ static struct irq_chip lan78xx_irqchip = {
 	.name			= "lan78xx-irqs",
 	.irq_mask		= lan78xx_irq_mask,
 	.irq_unmask		= lan78xx_irq_unmask,
+#ifdef CONFIG_IPIPE
+	.irq_hold = lan78xx_irq_mask,
+	.irq_release = lan78xx_irq_unmask,
+#endif
 	.irq_bus_lock		= lan78xx_irq_bus_lock,
 	.irq_bus_sync_unlock	= lan78xx_irq_bus_sync_unlock,
+	.flags			= IRQCHIP_PIPELINE_SAFE,
 };
 
 static int lan78xx_setup_irq_domain(struct lan78xx_net *dev)
-- 
2.17.1

