From 2e2e99a2a27b7ae87aafac2555f47802627e84d6 Mon Sep 17 00:00:00 2001
From: kknitin <nitin@elk.audio>
Date: Tue, 4 Feb 2020 22:37:47 +0000
Subject: [PATCH] PCIe irq-chip ipipe safe

---
 drivers/pci/controller/pcie-brcmstb.c | 4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

diff --git a/drivers/pci/controller/pcie-brcmstb.c b/drivers/pci/controller/pcie-brcmstb.c
index f99ab05ab477..0447a562353b 100644
--- a/drivers/pci/controller/pcie-brcmstb.c
+++ b/drivers/pci/controller/pcie-brcmstb.c
@@ -914,6 +914,7 @@ static struct irq_chip brcm_msi_irq_chip = {
 	.name = "Brcm_MSI",
 	.irq_mask = pci_msi_mask_irq,
 	.irq_unmask = pci_msi_unmask_irq,
+	.flags = IRQCHIP_PIPELINE_SAFE,
 };
 
 static struct msi_domain_info brcm_msi_domain_info = {
@@ -946,7 +947,7 @@ static void brcm_pcie_msi_isr(struct irq_desc *desc)
 			virq = irq_find_mapping(msi->inner_domain, hwirq);
 			if (virq) {
 				if (msi->used & (1 << hwirq))
-					generic_handle_irq(virq);
+					ipipe_handle_demuxed_irq(virq);
 				else
 					dev_info(dev, "unhandled MSI %d\n",
 						 hwirq);
@@ -982,6 +983,7 @@ static struct irq_chip brcm_msi_bottom_irq_chip = {
 	.name			= "Brcm_MSI",
 	.irq_compose_msi_msg	= brcm_compose_msi_msg,
 	.irq_set_affinity	= brcm_msi_set_affinity,
+	.flags			= IRQCHIP_PIPELINE_SAFE,
 };
 
 static int brcm_msi_alloc(struct brcm_msi *msi)
-- 
2.17.1

